convert_delta.1
.TH "convert_delta" "1" "Sept 25 2003" "" ""
.SH "NAME"
convert_delta
.SH "SYNOPSIS"
.PP
convert_delta -t FORMAT patch [other-patches] new-patch
.PP
.SH "DESCRIPTION"
convert_delta is a program for conversion of patch's from one format to 
another\&.  Also able to collapse multiple patches down to a single patch\&.
.SH "OPTIONS"
.PP
.nf

-V, --version                   output version, and exit
-v, --verbose                   increase verbosity
-c, --to-stdout                 output to-file to stdout
-s, --src-format FORMAT         used to override the auto
                                identification of a patch's format\&.
                                Valid formats are-
                                bdelta, gdiff4, gdiff5 (a nonstandard 
                                modification to gdiff4), bdiff, xdelta
                                and switching\&.
                                This option is not normally needed, 
                                since patcher can identify a patch's
                                format automatically if it is 
                                supported\&.
-t, --trg-format FORMAT         Used to specify the new patch's desired
                                format.
                                Valid formats are-
                                bdelta, gdiff4, gdiff5 (a nonstandard 
                                modification to gdiff4), bdiff, and
                                switching\&.
.fi
.PP
.SH "SEE ALSO"
.PP
differ(1)
diffball(1)
patcher(1)
.PP
.SH "BUGS"
none known
.PP
.SH "VERSION"
.PP
This man page is current for version 0\&.6\&.5 of convert_delta
.PP
.SH "CREDITS"
convert_delta, and by extension the diffball suite, is distributed under the GNU 
public license\&.  The file COPYING contains the details\&.
.PP
.SH "AUTHOR"
convert_delta was written by Brian Harring <ferringb@gentoo&.org>


NEWS
0.6.5
* Bug fixes.  bufferless reconstruction should work without issue now.
* Further code cleanup in preparation for attempting largefile (>2gb) support.
* convert_delta got a new feature; no longer limited to specifying only one patch for conversion.  You can now specify
  < 255 patches, and the patches are collapsed down into a single patch.  Doesn't do any optimization, but at this point 
  there really isn't any optimization done on the command lists.  Expect that sometime prior to 0.7

0.6.2 and under
* Reconstruction via multiple patches is now supported for all supported patch formats- no limitations on mixing formats also.
  I'd suspect a bug or two linger, but not many (been threw a fair amount of testing).
  
  No temps files used, each patch's command are merged into a single buffer in memory, rather then using intermediate files.

  Only limitation is number of registered dcb srcs; 256 is the current max.
  For switching/bdiff/bdelta/gdiff/xdelta, 256 patches are allowed.
  For bdiff, dependant on the version, figure 128- bsdiff is broken down into multiple bz2'd segments, so for reconstructing
  from a bsdiff patch patcher must maintain 2 decompressors into the file.

* Reading from compressed (gzip and bzip2, currently) patches/files is now supported, and pretty well tested.
  *NOTE* this is mainly for patcher, and convert_delta.  Using compressed srcs for differ/diffball isn't sane (due to issues 
  w/ seeking detailed below).

* During reconstruction, it's now possible to reorder commands so that file(s) used to reconstruct the target, are read linearly- 
  in general usage, this isn't needed unless src(s) are compressed, or multiple overlay-format patches are being applied 
  (ex: 2 bsdiff patches being applied).
  
  When a source is compressed, this makes a world of difference- gzip/bzip2 cseek implementations either decompress their 
  way forward to the desired location, or destruct the decompressor, re-init it, and read from the beginning of the cfile to the 
  desired position. 

  That behaviour sucks, but is something of the norm.  zlib's gzseek is implemented this way for instance.
  The commands are reordered such that copies from a src are in sequential order- writes to the reconstructed file are random.
  
  This isn't as quick as I'd like it to be, but its worst case is *much* saner then for dealing w/ seeking in compressed srcs.
  I would suspect it's performance could be tuned a bit, but currently it is better then the alternative.

* Default behaviour for applying single patches is now to _not_ read in the entire patch commands; now it just executes them as 
  it encounters them, unless A) applying multiple patches, B) decompressing files, C) reorder for read seq write random is in effect.
  Decreases memory usage nicely, and doesn't affect performance.

* One Half Pass Correcting alg implementation has been rewritten- basically the same, just minus an extra set of buffers.
  Quicker overall, and some long standing corner cases accessing beyond the buffers has been fixed.
  
* Lots of adjustments to CommandBuffer, and cfile.  Mostly expanding functionality, although there have been optimizations fixes.
  Introduced a DCB_registered_src struct- this is an abstraction to allow for overlay commands, and standard commands.
  Basically, things are saner/cleaner.  That, and a package wide reduction of intermediate buffers.


convert_delta.c
/*
  Copyright (C) 2003-2005 Brian Harring

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, US 
*/
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <stdio.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <string.h>
#include "cfile.h"
#include "dcbuffer.h"
#include "formats.h"
#include "defs.h"
#include "options.h"
#include "errors.h"

unsigned int global_verbosity = 0;

struct option long_opts[] = {
    STD_LONG_OPTIONS,
    FORMAT_LONG_OPTION("src-format",'s'),
    FORMAT_LONG_OPTION("trg-format",'t'),
    END_LONG_OPTS
};

struct usage_options help_opts[] = {
    STD_HELP_OPTIONS,
    FORMAT_HELP_OPTION("src-format", 's', "override auto-identification and specify source patches format"),
    FORMAT_HELP_OPTION("trg-format", 't', "override default and specify new patches format"),
    USAGE_FLUFF("convert_delta either expects 2 args (src patch, and new patches name), or just the source\n"
    "patches name if the option to dump to stdout has been given\n"
    "examples\n"
    "this would convert from the (auto-identified) xdelta format, to the default switching format\n"
    "convert_delta kde.xdelta kde.patch\n\n"
    "this would convert from the (auto-identified) xdelta format, to the gdiff4 format\n"
    "convert_delta kde.xdelta -t gdiff4 kde.patch\n"),
    END_HELP_OPTS
};

char short_opts[] = STD_SHORT_OPTIONS "s:t:";

int 
main(int argc, char **argv)
{
    int in_fh[256], out_fh;
    struct stat in_stat;
    CommandBuffer dcbuff[2];
    cfile in_cfh[256], out_cfh;
    char **patch_name;
    int optr, x;
    unsigned int patch_count;
    ECFH_ID src_id;
    signed int err;
    char *trg_file;
    unsigned long int src_format_id[256], trg_format_id=0;
    signed long recon_val=0, encode_result=0;
    unsigned int patch_compressor = 0;
    unsigned int output_to_stdout = 0;
    unsigned int global_use_md5 = 0;
    char *src_format = NULL, *trg_format = NULL;

    #define DUMP_USAGE(exit_code) \
	print_usage("convert_delta", "src_patch -t format [new_patch|or to stdout]", help_opts, exit_code)

    while((optr = getopt_long(argc, argv, short_opts, long_opts, NULL)) != -1) {
	switch(optr) {
	case OVERSION:
	    print_version("convert_delta");	exit(0);
	case OUSAGE:
	case OHELP:
	    DUMP_USAGE(0);
	case OVERBOSE:
	    global_verbosity++;
	    break;
	case 'f':
	    src_format = optarg;	break;
	case 't':
	    trg_format = optarg;	break;
	case OSTDOUT:
	    output_to_stdout = 1;	break;
	default:
	    v0printf("unknown option %s\n", argv[optind]);
	    DUMP_USAGE(EXIT_USAGE);

/*	case OBZIP2:
	    if(patch_compressor) {
		// bitch at em.
	    } else 
		patch_compressor = BZIP2_COMPRESSOR;
	    break;
	case OGZIP:
	    if(patch_compressor) {
		// bitch at em.
	    } else 
		patch_compressor = GZIP_COMPRESSOR;
	    break;
*/
	}
    }
/*    if( ((src_file=(char *)get_next_arg(argc, argv))==NULL) || 
	(stat(src_file, &in_stat))) {
	if(src_file) {
	    v0printf("Must specify an existing patch\n");
	    exit(EXIT_USAGE);
	}
	DUMP_USAGE(EXIT_USAGE);
    }
*/
    patch_count = argc - optind;
    patch_name = argv + optind;

    if(output_to_stdout) {
	out_fh = 1;
    } else {
	if(patch_count == 1) {
	    v0printf("you must specify at least a patch\n");
	    DUMP_USAGE(EXIT_USAGE);
	}
	trg_file = patch_name[patch_count - 1];
	if((out_fh = open(trg_file, O_WRONLY | O_TRUNC | O_CREAT, 0644))==-1){
	    v0printf( "error creating output file '%s'\n", trg_file);
	    exit(1);
	}
	patch_count--;
    }

    v1printf("patch_count is %u\n", patch_count);

    if(trg_format==NULL) {
	v0printf("new files format is required\n");
	DUMP_USAGE(EXIT_USAGE);
    } else {
	trg_format_id = check_for_format(trg_format, strlen(trg_format));
	if(trg_format_id==0) {
	    v0printf( "Unknown format '%s'\n", trg_format);
	    exit(1);
	}
    }

//    if(NULL!= get_next_arg(argc,argv)) {
//	DUMP_USAGE(EXIT_USAGE);
//    }

    if(src_format != NULL) {
	src_format_id[0] = check_for_format(src_format, strlen(src_format));
	if(src_format_id[0] == 0) {
	    v0printf("Unknown format '%s'\n", src_format);
	    exit(EXIT_FAILURE);
	}
	for(x=1; x < patch_count; x++) 
	    src_format_id[x] = src_format_id[0];
    }

    for(x=0; x < patch_count; x++) {
	v1printf("%u, opening %s\n", x, patch_name[x]);
	if((stat(patch_name[x], &in_stat)) || (in_fh[x] = open(patch_name[x], O_RDONLY, 0))==-1) {
	    v0printf( "error opening patch '%s', %d\n", patch_name[x], in_fh[x]);
	    exit(EXIT_FAILURE);
	}
	copen(in_cfh + x, in_fh[x], 0, in_stat.st_size, NO_COMPRESSOR, CFILE_RONLY);
	if(src_format == NULL) {
	    src_format_id[x] = identify_format(in_cfh + x);
	    if(src_format_id==0) {
		v0printf("Couldn't identify the patch format, aborting\n");
		exit(EXIT_FAILURE);
	    } else if((src_format_id[x] >> 16)==1) {
		v0printf( "Unsupported format version\n");
		exit(EXIT_FAILURE);
	    }
	    src_format_id[x] >>= 16;
	}
    }


//    DCB_full_init(&dcbuff, 4096,0,0);
    for(x = 0; x < patch_count; x++) {
	if(x == 0) {
	    err=DCB_full_init(&dcbuff[0], 4096, 0, 0);
	    check_return2(err, "DCBufferInit");
//	    src_id = internal_DCB_register_cfh_src(dcbuff, NULL, &bail_if_called_func, 
//		&bail_if_called_func, DC_COPY, 0);
	    src_id = DCB_register_fake_src(dcbuff, DC_COPY);
	    check_return2(src_id, "internal_DCB_register_cfh_src");
	} else {
	    err=DCB_full_init(&dcbuff[x % 2], 4096, 0,0);//dcbuff[(x - 1) % 2].ver_size, 0);
	    check_return2(err, "DCBufferInit");
	    src_id = DCB_register_dcb_src(&dcbuff[x % 2], &dcbuff[(x - 1) % 2]);
	    check_return2(src_id, "DCB_register_dcb_src");
	}

	if(SWITCHING_FORMAT == src_format_id[x]) {
	    recon_val = switchingReconstructDCBuff(src_id, &in_cfh[x], &dcbuff[x % 2]);
	} else if(GDIFF4_FORMAT == src_format_id[x]) {
            recon_val = gdiff4ReconstructDCBuff(src_id, &in_cfh[x], &dcbuff[x % 2]);
    	} else if(GDIFF5_FORMAT == src_format_id[x]) {
            recon_val = gdiff5ReconstructDCBuff(src_id, &in_cfh[x], &dcbuff[x % 2]);       
    	} else if(BDIFF_FORMAT == src_format_id[x]) {
            recon_val = bdiffReconstructDCBuff(src_id, &in_cfh[x], &dcbuff[x % 2]);       
    	} else if(XDELTA1_FORMAT == src_format_id[x]) {
            recon_val = xdelta1ReconstructDCBuff(src_id, &in_cfh[x], &dcbuff[x % 2], 1);
    	} else if(BDELTA_FORMAT == src_format_id[x]) {
            recon_val = bdeltaReconstructDCBuff(src_id, &in_cfh[x], &dcbuff[x % 2]);
    	} else if(BSDIFF_FORMAT == src_format_id[x]) {
	    v0printf("Sorry, unwilling to do bsdiff conversion in this version.\n");
	    v0printf("Try a newer version.\n");
	    exit(2);
//	    recon_val = bsdiffReconstructDCBuff(&in_cfh[x], &dcbuff[x % 2]);
//    	} else if(UDIFF_FORMAT == src_format_id[x]) {
//          recon_val = udiffReconstructDCBuff(&in_cfh[x], &src_cfh, NULL, &dcbuff[x % 2]);
	}
	v1printf("%u: resultant ver_size was %llu\n", x, (act_off_u64)dcbuff[x].ver_size);
	if(recon_val) {
	    v0printf("error detected while processing patch-quitting\n");
	    print_error(recon_val);
	    exit(EXIT_FAILURE);
	}
	if(x) {
	    DCBufferFree(&dcbuff[(x - 1) % 2]);
	}
    }

    v1printf("reconstruction return=%ld\n", recon_val);
    copen(&out_cfh, out_fh, 0, 0, NO_COMPRESSOR, CFILE_WONLY);
    v1printf("outputing patch...\n");
    if(DCBUFFER_FULL_TYPE == dcbuff[x % 2].DCBtype) {
	v1printf("there were %lu commands\n", ((DCB_full *)dcbuff[x % 2].DCB)->cl.com_count);
    }
    if(GDIFF4_FORMAT == trg_format_id) {
        encode_result = gdiff4EncodeDCBuffer(&dcbuff[(patch_count - 1) % 2], &out_cfh);
    } else if(GDIFF5_FORMAT == trg_format_id) {
        encode_result = gdiff5EncodeDCBuffer(&dcbuff[(patch_count - 1) % 2], &out_cfh);
    } else if(BDIFF_FORMAT == trg_format_id) {
        encode_result = bdiffEncodeDCBuffer(&dcbuff[(patch_count - 1) % 2], &out_cfh);
    } else if(SWITCHING_FORMAT == trg_format_id) {
        encode_result = switchingEncodeDCBuffer(&dcbuff[(patch_count - 1) % 2], &out_cfh);
    } else if (BDELTA_FORMAT == trg_format_id) {
        encode_result = bdeltaEncodeDCBuffer(&dcbuff[(patch_count - 1) % 2], &out_cfh);
    }
    v1printf("encoding return=%ld\n", encode_result);
    v1printf("finished.\n");
    DCBufferFree(&dcbuff[(patch_count -1) % 2]);
    for(x = 0; x < patch_count; x++) {
	cclose(&in_cfh[x]);
	close(in_fh[x]);
    }
    cclose(&out_cfh);
    return 0;
}
